chain/default_syncer.go

```diff
@@ -79,91 +89,74 @@ func NewDefaultSyncer(online, offline *hamt.CborIpldStore, c consensus.Protocol,
-		}
-		// try the network
-		if err = syncer.cstOnline.Get(ctx, blkCid, &blk); err != nil {
-			return nil, err
-		}
-		blks = append(blks, blk)
-	}
-	return blks, nil
+
+	return syncer.fetcher.GetBlocks(ctx, blkCids)
 }
 
-// collectChain resolves the cids of the head tipset and its ancestors to blocks
-// until it resolves blocks contained in the Store. collectChain may resolve cids
-// from the Store, the node's local offline cborstore, or the syncer's online
-// cbor store that is networked under the hood. collectChain errors if any
-// set of cids in the chain resolves to blocks that do not form a tipset, if
-// the chain is too long, or if any tipset has already been recorded as the
-// head of an invalid chain.
-//
-// collectChain is the entrypoint to the code that interacts with the network.
-// It does NOT add tipsets to the store.
-func (syncer *DefaultSyncer) collectChain(ctx context.Context, blkCids []cid.Cid) ([]types.TipSet, types.TipSet, error) {
+// collectChain resolves the cids of the head tipset and its ancestors to
+// blocks until it resolves a tipset with a parent contained in the Store. It
+// returns the chain of new incompletely validated tipsets and the id of the
+// parent tipset already synced into the store.  collectChain resolves cids
+// from the syncer's fetcher.  In production the fetcher wraps a bitswap
+// session.  collectChain errors if any set of cids in the chain resolves to
+// blocks that do not form a tipset, or if any tipset has already been recorded
+// as the head of an invalid chain.  collectChain is the entrypoint to the code
+// that interacts with the network. It does NOT add tipsets to the chainStore..
+func (syncer *DefaultSyncer) collectChain(ctx context.Context, blkCids []cid.Cid) ([]types.TipSet, error) {
 	var chain []types.TipSet
 	defer logSyncer.Info("chain synced")
 	for {
 		var blks []*types.Block
 		// check the cache for bad tipsets before doing anything
 		tsKey := types.NewSortedCidSet(blkCids...).String()
 
+		// Finish traversal if the tipset made is tracked in the store.
+		if syncer.chainStore.HasTipSetAndState(ctx, tsKey) {
+			return chain, nil
+		}
+
 		logSyncer.Debugf("CollectChain next link: %s", tsKey)
 
 		if syncer.badTipSets.Has(tsKey) {
-			return nil, nil, ErrChainHasBadTipSet
+			return nil, ErrChainHasBadTipSet
 		}
 
 		blks, err := syncer.getBlksMaybeFromNet(ctx, blkCids)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 
 		ts, err := syncer.consensus.NewValidTipSet(ctx, blks)
 		if err != nil {
 			syncer.badTipSets.Add(tsKey)
 			syncer.badTipSets.AddChain(chain)
-			return nil, nil, err
+			return nil, err
 		}
 
 		height, _ := ts.Height()
 		if len(chain)%500 == 0 {
 			logSyncer.Infof("syncing the chain, currently at block height %d", height)
-		}
-
-		// Finish traversal if the tipset made is tracked in the store.
-		if syncer.chainStore.HasTipSetAndState(ctx, tsKey) {
-			return chain, ts, nil
+			syncer.fetcher.LogStat()
```

Comment:
i'd lean toward making this method actually return the stats, and then print it where needed. Me putting this here was really just a debug log to help observe the improvements.

ipfs has this wired through as a command.

Reply:
Because the blockservice is needed to get the stats and we are talking about removing this from the Fetcher I am just going to remove logstats altogether and file a followon issue for wiring bitswap up to a command for viewing stats.